packages/apollo-server-core/src/ApolloServer.ts

```diff
@@ -449,9 +519,19 @@ export class ApolloServerBase {
       path,
     } = this.subscriptionServerOptions;
 
+    // TODO: This shouldn't use the schema hack.
+    //
+    // If this method can be made async we can remove this hack in favor of
+    // grabbing the schema from the createSchemaDerivedData promise
```

Comment:
Making this an `async` function would be a breaking change which we can't afford just yet (Apollo Server 3 will afford us this opportunity and will completely change the way that the subscription server behaves; AS3 should make subscriptions a first-class citizen).

Since `gateway` is a new feature, I think we are willing to say, "`subscriptions` must be `false` when `gateway` is enabled.".

What if, rather than setting `__hackedSchema` in the asynchronous `initSchema`, we moved anything which doesn't need asynchronicity (the latter half of that method, I think?) into a non-`async` method and, in the `constructor`, if `!gateway && schema` we just set `__hackedSchema` to `schema` (with a `XXX Here be subscription demons.` note, probably) and let this `installSubscriptionsHandler` remain synchronous?  

It's worth pointing out that the subscriptions server makes no benefit of `schemaHash`, `documentStore` or `extensions`, so `schema` is the sole concern — though we should keep providing a schema with the appropriate "`augmentedTypeDefs`" and mock functions to ensure compatibility with existing implementors.

Thoughts?

Reply:
The current default is `subscriptions` are enabled unless explicitly set `false`. I think we can move that to "`subscriptions` are enabled unless explicitly set `false`, or `gateway` is defined. If `gateway` is defined and `subscriptions` is truthy, an error is thrown."

Moved to `initSchema` returning a `GraphQLSchema | Promise<GraphQLSchema` and working from there. It's not as pretty but it ensures `this.schema` is synchronously defined.